tedana/utils/utils.py

```diff
@@ -1,198 +1,452 @@
-
-        mthr = np.ones([nx, ny, nz, Ne])
-        for i_echo in range(Ne):
-            mthr[:, :, :, i_echo] *= lthrs[i_echo]
-        mthr = np.abs(emeans) > mthr
-        masksum = np.array(mthr, dtype=np.int).sum(-1)
-        mask = masksum != 0
-        if getsum:
-            return mask, masksum
-        else:
-            return mask
+    fdata = data.reshape((-1,) + data.shape[3:])
+
+    return fdata.squeeze()
 
 
-def uncat2echos(data, Ne):
+def load_data(data, n_echos=None):
     """
-    Combines z- and echo-axis in `data`
+    Coerces input `data` files to required 3D array output
 
     Parameters
     ----------
-    data : array_like
-        Array of shape (nx, ny, nz, Ne, nt)
-    Ne : int
-        Number of echoes; should be equal to `data.shape[3]`
+    data : (X x Y x M x T) array_like or list-of-img_like
+        Input multi-echo data array, where `X` and `Y` are spatial dimensions,
+        `M` is the Z-spatial dimensions with all the input echos concatenated,
+        and `T` is time. A list of image-like objects (e.g., .nii or .gii) are
+        accepted, as well
+    n_echos : int, optional
+        Number of echos in provided data array. Only necessary if `data` is
+        array_like. Default: None
 
     Returns
     -------
-    ndarray
-        Array of shape (nx, ny, nz*Ne, nt)
+    fdata : (S x E x T) np.ndarray
+        Output data where `S` is samples, `E` is echos, and `T` is time
+    ref_img : str
+        Filepath to reference image for saving output files
     """
 
-    nx, ny = data.shape[0:2]
-    nz = data.shape[2] * Ne
-    if len(data.shape) > 4:
-        nt = data.shape[4]
-    else:
-        nt = 1
-    return np.reshape(data, (nx, ny, nz, nt), order='F')
+    if isinstance(data, list):
+        if get_dtype(data) == 'GIFTI':  # TODO: deal with L/R split GIFTI files
+            pass
+        if len(data) == 1:  # a z-concatenated file was provided
+            if n_echos is None:
+                raise ValueError('Number of echos `n_echos` must be specified '
+                                 'if z-concatenated data file provided.')
+            data = data[0]
+        elif len(data) == 2:  # inviable -- need more than 2 echos
+            raise ValueError('Cannot run `tedana` with only two echos: '
+                             '{}'.format(data))
+        else:  # individual echo files were provided
+            fdata = np.stack([load_image(f) for f in data], axis=1)
+            return np.atleast_3d(fdata), data[0]
+
+    # we have a z-cat file
+    img = check_niimg(data)
+    (nx, ny), nz = img.shape[:2], img.shape[2] // n_echos
+    fdata = load_image(img.get_data().reshape(nx, ny, nz, n_echos, -1))
+
+    # create reference image
+    ref_img = img.__class__(np.zeros((nx, ny, nz)), affine=img.affine,
+                            header=img.header, extra=img.extra)
+
+    return fdata, ref_img
+
+
+def makeadmask(data, minimum=True, getsum=False):
```

Comment:
Should we rename this? I'm still not sure what `ad` in `makeadmask` stands for.

Reply:
Definitely can be renamed. I believe it stood for **make** **ad**apative **mask**, but since that's clear to exactly no one, a more pythonic naming convention (`make_adaptive_mask()`) might be more applicable? I'm happy to do whatever works best!